AEMG/data_utils.py

```diff
@@ -95,12 +95,72 @@ def __init__(self,config):
         self.labels = np.array(self.labels)
         self.labels = torch.from_numpy(self.labels).long()
 
+        self.generate_contrastive_triplets()
+
     def __len__(self):
-        return len(self.final_points)
+        return len(self.contrastive_triplets)
 
     def __getitem__(self, idx):
-        return self.final_points[idx], self.labels[idx]         
-            
+        return self.contrastive_triplets[idx]
+
+    def generate_contrastive_triplets(self):
+        """
+        Generates the contrastive triplets for the dataset where each triplet has the form (anchor, positive, negative)
+        The anchor can be either a success or failure point, the positive is a point of the same class as the anchor, and the negative is a point of the opposite class
+        """
+
+        # Gathering the indices of the success and failure labels
+        success_indices = torch.where(self.labels == 1)[0]
+        failure_indices = torch.where(self.labels == 0)[0]
+
+        # Generating all possible combinations of success and failure indices for positive anchor pairs
+        success_anchor_pairs = torch.stack(torch.meshgrid(success_indices, success_indices)).T.reshape(-1,2)
+        failure_anchor_pairs = torch.stack(torch.meshgrid(failure_indices, failure_indices)).T.reshape(-1,2)
+
+        # Removing instances with the same index
+        success_anchor_pairs = success_anchor_pairs[success_anchor_pairs[:, 0] != success_anchor_pairs[:, 1]]
+        failure_anchor_pairs = failure_anchor_pairs[failure_anchor_pairs[:, 0] != failure_anchor_pairs[:, 1]]
+
+        success_indices = success_indices[:, None]
+        failure_indices = failure_indices[:, None]
+
+        # Generating the sample size for the negative anchor pairs
+        success_failure_sample_size = 10
+        num_success_failure_triplets = len(success_anchor_pairs) * success_failure_sample_size
+        failure_success_sample_size = min(num_success_failure_triplets // len(failure_anchor_pairs), len(success_indices))
+
+        # Adding negative values to the positive anchor pairs
+        success_failure_triplets = multi_dim_tensor_cartesian(success_anchor_pairs, failure_indices[torch.randint(0, len(failure_indices), (success_failure_sample_size, ))])
+        failure_success_triplets = multi_dim_tensor_cartesian(failure_anchor_pairs, success_indices[torch.randint(0, len(success_indices), (failure_success_sample_size, ))])
+
+        # Concatenating all triplets
+        self.contrastive_triplets = torch.cat([success_failure_triplets, failure_success_triplets], dim=0)
+
+    def collate_fn(self, batch):
```

Comment:
If we make `self.contrastive_triplets` itself a `Dict,` would we not need this function?

Reply:
We will still need it to collect the batch of final points. Right now, it collects only the final points that we need in the current batch and not all available final points. I'm not sure if this operation will result in slowness though